File: runlens/efforts.py
"""Layer 0 data hygiene + the Steady Effort Store (PROPOSAL.md v2).

Pure functions. Everything the efficiency statistics read comes from
here — nothing downstream reads raw session averages or unfiltered HR.

Hygiene (optical wrist-HR failure modes):
- spikes: |ΔHR| > 8 bpm between consecutive 1 Hz samples → interpolate
  gaps ≤ 5 s, flag longer stretches invalid
- cadence lock: |HR − cadence(spm)| < 3 for ≥ 30 s → flagged suspect
- stabilized HR is a median, never a mean

Efforts:
- reps (engine bounds): pace always; HR only stabilized (first 45 s
  dropped, median of the rest); reps < 90 s contribute no HR statistics
- steady blocks: the progression rule (≥ 600 s, pace CV ≤ 5%, no pause)
- everything else (warmup/cooldown/recovery/junk) never aggregates

Domains (percent of LTHR, so recalibration moves all boundaries):
  easy ≤ Z2 ceiling < steady < 93% LTHR ≤ threshold ≤ 102% LTHR < hard
"""

from __future__ import annotations

import math
import statistics
from datetime import timedelta

HR_SPIKE_DELTA = 8          # bpm between consecutive 1 Hz samples
HR_GAP_INTERPOLATE_S = 5
CADENCE_LOCK_TOL = 3        # bpm vs steps/min
CADENCE_LOCK_MIN_S = 30
HR_STABILIZE_S = 45         # HR time constant ~30-45 s
HR_MIN_REP_S = 90           # below this a rep has no stabilized HR
HR_MIN_SAMPLES = 20

DEFAULT_SETTINGS = {
    "lthr": 176.0,           # @ 3:55/km (Garmin, to confirm)
    "max_hr": 190.0,
    "z2_ceiling": 143.0,
    "ref_hr": 175.0,         # LTHR - 1
    "lthr_pace_s_km": 235.0, # 3:55/km, for the pace-based fallback
    "boundary_version": 1,
}

DOMAINS = ("easy", "steady", "threshold", "hard")


# ------------------------------------------------------------ hygiene

def clean_heart_rate(records: list[dict]) -> list[dict]:
    """Return records with an added 'hr_ok' flag and spike-interpolated
    HR. Original list is not mutated."""
    out = [dict(r) for r in records]

    # Spikes / dropouts: physiologically impossible deltas.
    last_good = None
    bad_run: list[int] = []
    for i, r in enumerate(out):
        hr = r.get("heart_rate")
        ok = hr is not None and 30 <= hr <= 230
        if ok and last_good is not None:
            prev_hr, prev_i = last_good
            dt = max(1, (out[i]["timestamp"] - out[prev_i]["timestamp"]).total_seconds())
            if abs(hr - prev_hr) > HR_SPIKE_DELTA * dt:
                ok = False
        if ok:
            if bad_run and last_good is not None:
                prev_hr, prev_i = last_good
                gap = (out[i]["timestamp"] - out[prev_i]["timestamp"]).total_seconds()
                if gap <= HR_GAP_INTERPOLATE_S:
                    for k, j in enumerate(bad_run, 1):
                        frac = k / (len(bad_run) + 1)
                        out[j]["heart_rate"] = round(prev_hr + frac * (hr - prev_hr))
                        out[j]["hr_ok"] = True
            bad_run = []
            last_good = (hr, i)
            r["hr_ok"] = True
        else:
            r.setdefault("hr_ok", False)
            bad_run.append(i)

    # Cadence lock: HR glued to steps/min for a sustained stretch.
    run_start = None
    for i, r in enumerate(out + [None]):
        locked = (r is not None and r.get("hr_ok")
                  and r.get("heart_rate") and r.get("cadence")
                  and abs(r["heart_rate"] - r["cadence"]) < CADENCE_LOCK_TOL)
        if locked and run_start is None:
            run_start = i
        elif not locked and run_start is not None:
            span = (out[i - 1]["timestamp"] - out[run_start]["timestamp"]).total_seconds()
            if span >= CADENCE_LOCK_MIN_S:
                for j in range(run_start, i):
                    out[j]["hr_ok"] = False
                    out[j]["hr_suspect"] = True
            run_start = None
    return out


def smooth_speed(records: list[dict], window: int = 7) -> list[dict]:
    """Rolling-mean speed (GPS pace is noise at 1 Hz). Adds 'speed_smooth'."""
    speeds = [r.get("enhanced_speed") for r in records]
    half = window // 2
    for i, r in enumerate(records):
        chunk = [s for s in speeds[max(0, i - half):i + half + 1] if s is not None]
        r["speed_smooth"] = statistics.mean(chunk) if chunk else None
    return records


# ------------------------------------------------------------ domains

def domain_boundaries(settings: dict) -> dict:
    lthr = settings["lthr"]
    return {
        "easy_max": settings["z2_ceiling"],
        "steady_max": 0.93 * lthr,
        "threshold_max": 1.02 * lthr,
    }


def domain_for_hr(hr: float, settings: dict) -> str:
    b = domain_boundaries(settings)
    if hr <= b["easy_max"]:
        return "easy"
    if hr < b["steady_max"]:
        return "steady"
    if hr <= b["threshold_max"]:
        return "threshold"
    return "hard"


def domain_for_pace(pace_s_km: float, settings: dict) -> str:
    """Pace-based fallback (reps < 90 s, hr_suspect): bands derived from
    LTHR pace. Rough by design — pace-only statistics."""
    ltp = settings["lthr_pace_s_km"]
    if pace_s_km < 0.95 * ltp:
        return "hard"
    if pace_s_km <= 1.05 * ltp:
        return "threshold"
    if pace_s_km <= 1.15 * ltp:
        return "steady"
    return "easy"


# ------------------------------------------------------------ extraction

def _records_between(records, start, end):
    return [r for r in records if start <= r["timestamp"] <= end]


def _stabilized_hr(records: list[dict], start, end) -> tuple[float | None, bool, int]:
    """(median stabilized HR, suspect_majority, n_samples)."""
    window = _records_between(records, start + timedelta(seconds=HR_STABILIZE_S), end)
    good = [r["heart_rate"] for r in window if r.get("hr_ok")]
    suspect = sum(1 for r in window if r.get("hr_suspect"))
    if len(good) < HR_MIN_SAMPLES:
        return None, suspect > len(window) / 2 if window else False, len(good)
    return float(statistics.median(good)), suspect > len(window) / 2, len(good)


def extract_efforts(segments: list[dict], records: list[dict],
                    pauses: list[tuple], settings: dict | None = None) -> list[dict]:
    """The Steady Effort Store for one session.

    records must have passed clean_heart_rate() and smooth_speed().
    Returns effort dicts: kind rep|steady, pace, stabilized HR (or None),
    hr_valid, domain (+basis), REI, avg power.
    """
    from . import metrics as _m
    settings = {**DEFAULT_SETTINGS, **(settings or {})}
    efforts: list[dict] = []

    reps = [s for s in segments if s["kind"] == "rep" and not s.get("outlier")]
    for seg in reps:
        start, end = seg.get("start_time"), seg.get("end_time")
        timer = seg.get("timer_s") or 0
        if not start or not end or not timer:
            continue
        speed = seg.get("avg_speed_ms") or (
            (seg.get("distance_m") or 0) / timer if timer else None)
        if not speed:
            continue
        pace = 1000.0 / speed
        hr_med, suspect, n_hr = (None, False, 0)
        if timer >= HR_MIN_REP_S:
            hr_med, suspect, n_hr = _stabilized_hr(records, start, end)
        hr_valid = hr_med is not None and not suspect
        window = _records_between(records, start, end)
        powers = [r["power"] for r in window if r.get("power")]
        if hr_valid:
            dom, basis = domain_for_hr(hr_med, settings), "hr"
        else:
            dom, basis = domain_for_pace(pace, settings), "pace"
        efforts.append({
            "kind": "rep",
            "rep_index": seg.get("rep_index"),
            "start_time": start, "end_time": end,
            "timer_s": timer, "distance_m": seg.get("distance_m"),
            "speed_ms": speed, "pace_s_km": round(pace, 1),
            "hr_median": round(hr_med, 1) if hr_med is not None else None,
            "hr_valid": hr_valid, "hr_suspect": suspect,
            "domain": dom, "domain_basis": basis,
            "rei": round(speed / hr_med, 5) if hr_valid else None,
            "avg_power": round(statistics.mean(powers), 1) if powers else None,
            "boundary_version": settings["boundary_version"],
        })

    # Steady continuous blocks (never from interval sessions' recoveries;
    # the progression rule already guarantees that).
    for cand in _m.find_progression_candidates(segments, records, pauses):
        start, end = cand["start_time"], cand["end_time"]
        hr_med, suspect, _ = _stabilized_hr(records, start, end)
        hr_valid = hr_med is not None and not suspect
        window = _records_between(records, start, end)
        powers = [r["power"] for r in window if r.get("power")]
        speed = 1000.0 / cand["pace_s_km"]
        if hr_valid:
            dom, basis = domain_for_hr(hr_med, settings), "hr"
        else:
            dom, basis = domain_for_pace(cand["pace_s_km"], settings), "pace"
        efforts.append({
            "kind": "steady",
            "rep_index": None,
            "start_time": start, "end_time": end,
            "timer_s": cand["timer_s"], "distance_m": None,
            "speed_ms": round(speed, 3), "pace_s_km": cand["pace_s_km"],
            "hr_median": round(hr_med, 1) if hr_med is not None else None,
            "hr_valid": hr_valid, "hr_suspect": suspect,
            "domain": dom, "domain_basis": basis,
            "rei": round(speed / hr_med, 5) if hr_valid else None,
            "avg_power": round(statistics.mean(powers), 1) if powers else None,
            "boundary_version": settings["boundary_version"],
        })
    return efforts

File: docs/js/efforts.js
// Layer 0 hygiene + Steady Effort Store — 1:1 port of runlens/efforts.py.

import { findProgressionCandidates } from './metrics.js';

export const HR_SPIKE_DELTA = 8;
export const HR_GAP_INTERPOLATE_S = 5;
export const CADENCE_LOCK_TOL = 3;
export const CADENCE_LOCK_MIN_S = 30;
export const HR_STABILIZE_S = 45;
export const HR_MIN_REP_S = 90;
export const HR_MIN_SAMPLES = 20;

export const DEFAULT_SETTINGS = {
  lthr: 176.0, max_hr: 190.0, z2_ceiling: 143.0, ref_hr: 175.0,
  lthr_pace_s_km: 235.0, boundary_version: 1,
};
export const DOMAINS = ['easy', 'steady', 'threshold', 'hard'];

const median = (xs) => {
  const s = [...xs].sort((a, b) => a - b);
  const n = s.length;
  return n ? (n % 2 ? s[(n - 1) / 2] : (s[n / 2 - 1] + s[n / 2]) / 2) : null;
};
const mean = (xs) => xs.length ? xs.reduce((a, b) => a + b, 0) / xs.length : null;

export function cleanHeartRate(records) {
  const out = records.map(r => ({ ...r }));
  let lastGood = null;
  let badRun = [];
  for (let i = 0; i < out.length; i++) {
    const r = out[i];
    const hr = r.heart_rate;
    let ok = hr !== null && hr !== undefined && hr >= 30 && hr <= 230;
    if (ok && lastGood) {
      const [prevHr, prevI] = lastGood;
      const dt = Math.max(1, (out[i].timestamp - out[prevI].timestamp) / 1000);
      if (Math.abs(hr - prevHr) > HR_SPIKE_DELTA * dt) ok = false;
    }
    if (ok) {
      if (badRun.length && lastGood) {
        const [prevHr, prevI] = lastGood;
        const gap = (out[i].timestamp - out[prevI].timestamp) / 1000;
        if (gap <= HR_GAP_INTERPOLATE_S) {
          badRun.forEach((j, k) => {
            const frac = (k + 1) / (badRun.length + 1);
            out[j].heart_rate = Math.round(prevHr + frac * (hr - prevHr));
            out[j].hr_ok = true;
          });
        }
      }
      badRun = [];
      lastGood = [hr, i];
      r.hr_ok = true;
    } else {
      if (r.hr_ok === undefined) r.hr_ok = false;
      badRun.push(i);
    }
  }
  // Cadence lock detection.
  let runStart = null;
  for (let i = 0; i <= out.length; i++) {
    const r = out[i] ?? null;
    const locked = r && r.hr_ok && r.heart_rate && r.cadence &&
      Math.abs(r.heart_rate - r.cadence) < CADENCE_LOCK_TOL;
    if (locked && runStart === null) runStart = i;
    else if (!locked && runStart !== null) {
      const span = (out[i - 1].timestamp - out[runStart].timestamp) / 1000;
      if (span >= CADENCE_LOCK_MIN_S) {
        for (let j = runStart; j < i; j++) {
          out[j].hr_ok = false;
          out[j].hr_suspect = true;
        }
      }
      runStart = null;
    }
  }
  return out;
}

export function smoothSpeed(records, window = 7) {
  const speeds = records.map(r => r.enhanced_speed);
  const half = Math.floor(window / 2);
  records.forEach((r, i) => {
    const chunk = speeds.slice(Math.max(0, i - half), i + half + 1)
      .filter(s => s !== null && s !== undefined);
    r.speed_smooth = chunk.length ? mean(chunk) : null;
  });
  return records;
}

export function domainBoundaries(settings) {
  return { easy_max: settings.z2_ceiling,
           steady_max: 0.93 * settings.lthr,
           threshold_max: 1.02 * settings.lthr };
}

export function domainForHr(hr, settings) {
  const b = domainBoundaries(settings);
  if (hr <= b.easy_max) return 'easy';
  if (hr < b.steady_max) return 'steady';
  if (hr <= b.threshold_max) return 'threshold';
  return 'hard';
}

export function domainForPace(pace, settings) {
  const ltp = settings.lthr_pace_s_km;
  if (pace < 0.95 * ltp) return 'hard';
  if (pace <= 1.05 * ltp) return 'threshold';
  if (pace <= 1.15 * ltp) return 'steady';
  return 'easy';
}

function recordsBetween(records, start, end) {
  return records.filter(r => r.timestamp >= start && r.timestamp <= end);
}

function stabilizedHr(records, start, end) {
  const from = new Date(start.getTime() + HR_STABILIZE_S * 1000);
  const window = recordsBetween(records, from, end);
  const good = window.filter(r => r.hr_ok).map(r => r.heart_rate);
  const suspect = window.filter(r => r.hr_suspect).length;
  const suspectMaj = window.length ? suspect > window.length / 2 : false;
  if (good.length < HR_MIN_SAMPLES) return [null, suspectMaj, good.length];
  return [median(good), suspectMaj, good.length];
}

export function extractEfforts(segments, records, pauses, settings = null) {
  const s = { ...DEFAULT_SETTINGS, ...(settings || {}) };
  const efforts = [];

  const reps = segments.filter(x => x.kind === 'rep' && !x.outlier);
  for (const seg of reps) {
    const start = seg.start_time, end = seg.end_time;
    const timer = seg.timer_s ?? 0;
    if (!start || !end || !timer) continue;
    const speed = seg.avg_speed_ms ?? (timer ? (seg.distance_m ?? 0) / timer : null);
    if (!speed) continue;
    const pace = 1000 / speed;
    let [hrMed, suspect] = [null, false];
    if (timer >= HR_MIN_REP_S) {
      [hrMed, suspect] = stabilizedHr(records, start, end);
    }
    const hrValid = hrMed !== null && !suspect;
    const window = recordsBetween(records, start, end);
    const powers = window.map(r => r.power).filter(Boolean);
    efforts.push({
      kind: 'rep', rep_index: seg.rep_index ?? null,
      start_time: start, end_time: end,
      timer_s: timer, distance_m: seg.distance_m ?? null,
      speed_ms: speed, pace_s_km: Math.round(pace * 10) / 10,
      hr_median: hrMed !== null ? Math.round(hrMed * 10) / 10 : null,
      hr_valid: hrValid, hr_suspect: suspect,
      domain: hrValid ? domainForHr(hrMed, s) : domainForPace(pace, s),
      domain_basis: hrValid ? 'hr' : 'pace',
      rei: hrValid ? speed / hrMed : null,
      avg_power: powers.length ? Math.round(mean(powers) * 10) / 10 : null,
      boundary_version: s.boundary_version,
    });
  }

  for (const cand of findProgressionCandidates(segments, records, pauses)) {
    const [hrMed, suspect] = stabilizedHr(records, cand.start_time, cand.end_time);
    const hrValid = hrMed !== null && !suspect;
    const window = recordsBetween(records, cand.start_time, cand.end_time);
    const powers = window.map(r => r.power).filter(Boolean);
    const speed = 1000 / cand.pace_s_km;
    efforts.push({
      kind: 'steady', rep_index: null,
      start_time: cand.start_time, end_time: cand.end_time,
      timer_s: cand.timer_s, distance_m: null,
      speed_ms: Math.round(speed * 1000) / 1000, pace_s_km: cand.pace_s_km,
      hr_median: hrMed !== null ? Math.round(hrMed * 10) / 10 : null,
      hr_valid: hrValid, hr_suspect: suspect,
      domain: hrValid ? domainForHr(hrMed, s) : domainForPace(cand.pace_s_km, s),
      domain_basis: hrValid ? 'hr' : 'pace',
      rei: hrValid ? speed / hrMed : null,
      avg_power: powers.length ? Math.round(mean(powers) * 10) / 10 : null,
      boundary_version: s.boundary_version,
    });
  }
  return efforts;
}
